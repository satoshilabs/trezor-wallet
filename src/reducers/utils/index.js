/* @flow */
import BigNumber from 'bignumber.js';

import type {
    State,
    Device,
    TrezorDevice,
    Account,
    Coin,
    Discovery,
    Token,
    PendingTx,
    Web3Instance,
} from 'flowtype';

export const getSelectedDevice = (state: State): ?TrezorDevice => {
    const locationState = state.router.location.state;
    if (!locationState.device) return undefined;

    const instance: ?number = locationState.deviceInstance ? parseInt(locationState.deviceInstance, 10) : undefined;
    return state.devices.find((d) => {
        if (!d.features && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.bootloader_mode && d.path === locationState.device) {
            return true;
        } if (d.features && d.features.device_id === locationState.device && d.instance === instance) {
            return true;
        }
        return false;
    });
};

//
export const isSelectedDevice = (current: ?TrezorDevice, device: ?TrezorDevice): boolean => !!((current && device && (current.path === device.path && current.instance === device.instance)));

// find device by id and state
export const findDevice = (devices: Array<TrezorDevice>, deviceId: string, deviceState: string /*, instance: ?number*/): ?TrezorDevice => devices.find((d) => {
    // TODO: && (instance && d.instance === instance)
    if (d.features && d.features.device_id === deviceId && d.state === deviceState) {
        return true;
    }
    return false;
});

// get next instance number
export const getDuplicateInstanceNumber = (devices: Array<TrezorDevice>, device: Device | TrezorDevice): number => {
    // find device(s) with the same features.device_id
    // and sort them by instance number
    const affectedDevices: Array<TrezorDevice> = devices.filter(d => d.features && device.features && d.features.device_id === device.features.device_id)
        .sort((a, b) => {
            if (!a.instance) {
                return -1;
            }
            return !b.instance || a.instance > b.instance ? 1 : -1;
        });

    // calculate new instance number
    const instance: number = affectedDevices.reduce((inst, dev) => (dev.instance ? dev.instance + 1 : inst + 1), 0);
    return instance;
};

export const getSelectedAccount = (state: State): ?Account => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network || !locationState.account) return null;

    const index: number = parseInt(locationState.account, 10);

    return state.accounts.find(a => a.deviceState === device.state && a.index === index && a.network === locationState.network);
};

export const getSelectedNetwork = (state: State): ?Coin => {
    const device = state.wallet.selectedDevice;
    const { coins } = state.localStorage.config;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return coins.find(c => c.network === locationState.network);
};

export const getDiscoveryProcess = (state: State): ?Discovery => {
    const device = state.wallet.selectedDevice;
    const locationState = state.router.location.state;
    if (!device || !locationState.network) return null;

    return state.discovery.find(d => d.deviceState === device.state && d.network === locationState.network);
};

export const getAccountPendingTx = (pending: Array<PendingTx>, account: ?Account): Array<PendingTx> => {
    const a = account;
    if (!a) return [];
    return pending.filter(p => p.network === a.network && p.address === a.address);
};

export const getPendingNonce = (pending: Array<PendingTx>): number => pending.reduce((value: number, tx: PendingTx) => {
    if (tx.rejected) return value;
    return Math.max(value, tx.nonce + 1);
}, 0);

export const getPendingAmount = (pending: Array<PendingTx>, currency: string, token: boolean = false): BigNumber => pending.reduce((value: BigNumber, tx: PendingTx) => {
    if (tx.currency === currency && !tx.rejected) {
        return new BigNumber(value).plus(token ? tx.amount : tx.total);
    }
    return value;
}, new BigNumber('0'));

export const getAccountTokens = (state: State, account: ?Account): Array<Token> => {
    const a = account;
    if (!a) return [];
    return state.tokens.filter(t => t.ethAddress === a.address && t.network === a.network && t.deviceState === a.deviceState);
};

export const getWeb3 = (state: State): ?Web3Instance => {
    const locationState = state.router.location.state;
    if (!locationState.network) return null;
    return state.web3.find(w3 => w3.network === locationState.network);
};

export const observeChanges = (prev: ?Object, current: ?Object, filter?: {[k: string]: Array<string>}): boolean => {
    // 1. both objects are the same (solves simple types like string, boolean and number)
    if (prev === current) return false;
    // 2. one of the objects is null/undefined
    if (!prev || !current) return true;

    const prevType = Object.prototype.toString.call(prev);
    const currentType = Object.prototype.toString.call(current);
    // 3. one of the objects has different type then other
    if (prevType !== currentType) return true;

    if (currentType === '[object Array]') {
        // 4. Array length is different
        if (prev.length !== current.length) return true;
        // observe array recursive
        for (let i = 0; i < current.length; i++) {
            if (observeChanges(prev[i], current[i], filter)) return true;
        }
    } else if (currentType === '[object Object]') {
        const prevKeys = Object.keys(prev);
        const currentKeys = Object.keys(current);
        // 5. simple validation of keys length
        if (prevKeys.length !== currentKeys.length) return true;

        // 6. "prev" has keys which "current" doesn't have
        const prevDifference = prevKeys.find(k => currentKeys.indexOf(k) < 0);
        if (prevDifference) return true;

        // 8. observe every key recursive
        for (let i = 0; i < currentKeys.length; i++) {
            const key = currentKeys[i];
            if (filter && filter.hasOwnProperty(key) && prev[key] && current[key]) {
                const prevFiltered = {};
                const currentFiltered = {};
                for (let i2 = 0; i2 < filter[key].length; i2++) {
                    const field = filter[key][i2];
                    prevFiltered[field] = prev[key][field];
                    currentFiltered[field] = current[key][field];
                }
                if (observeChanges(prevFiltered, currentFiltered)) return true;
            } else if (observeChanges(prev[key], current[key])) {
                return true;
            }
        }
    } else if (prev !== current) {
        // solve simple types like string, boolean and number
        return true;
    }

    return false;
};
