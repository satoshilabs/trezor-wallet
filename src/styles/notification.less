.notification {
    position: relative;
    color: @color_info_primary;
    background: @color_info_secondary;
    padding: 24px 48px 24px 80px;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    text-align: left;

    .notification-body {
        flex: 1;
        margin-right: 24px;
    }

    .notification-action button {
        padding: 12px 58px;
    }

    .notification-close {
        position: absolute;
        top: 8px;
        right: 0;
        padding: 12px;
        color: inherit;
        transition: opacity 0.3s;
        z-index: 1;

        &:after {
            .icomoon-close;
        }
        &:active,
        &:hover {
            opacity: 0.6;
            color: inherit;
        }
    }

    h2 {
        font-size: 14px;
        font-weight: bold;

        padding: 0px;
        &:before {
            .icomoon-info;
            position: absolute;
            top: 17px;
            left: 40px;
            font-size: 32px !important;
        }
    }

    p {
        padding: 0px;
        margin: 8px 0px;
        color: inherit;
    }

    &.info {
        .notification-action button {
            border: 1px solid @color_info_primary;
            color: @color_info_primary;
            &:hover {
                color: @color_white;
                background: @color_info_primary;
            }
        }
    }



    &.success {
        color: @color_success_primary;
        background: @color_success_secondary;

        .notification-action button {
            border: 1px solid @color_success_primary;
            color: @color_success_primary;
            &:hover {
                color: @color_white;
                background: @color_success_primary;
            }
        }
    }

    &.warning {
        color: @color_warning_primary;
        background: @color_warning_secondary;
        h2:before {
            .icomoon-warning;
        }

        .notification-action button {
            border: 1px solid @color_warning_primary;
            color: @color_warning_primary;
            &:hover {
                color: @color_white;
                background: @color_warning_primary;
            }
        }
    }

    &.error {
        color: @color_error_primary;
        background: @color_error_secondary;
        h2:before {
            .icomoon-error;
        }

        .notification-close {
            color: @color_error_primary;
            &:hover {
                color: @color_error_primary;
            }
        }

        .notification-action button {
            border: 1px solid @color_error_primary;
            color: @color_error_primary;
            &:hover {
                color: @color_white;
                background: @color_error_primary;
            }
        }
    }
}