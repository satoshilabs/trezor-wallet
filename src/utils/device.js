/* @flow */

import colors from 'config/colors';

import type { Device } from 'trezor-connect';
import type { TrezorDevice, State } from 'flowtype';

type Transport = $ElementType<$ElementType<State, 'connect'>, 'transport'>;

export const getStatus = (device: TrezorDevice): string => {
    if (!device.connected) {
        return 'disconnected';
    }
    if (device.type === 'acquired') {
        if (device.mode === 'bootloader') {
            return 'bootloader';
        }
        if (device.mode === 'initialize') {
            return 'initialize';
        }
        if (device.mode === 'seedless') {
            return 'seedless';
        }
        if (device.firmware === 'required') {
            return 'firmware-required';
        }
        if (device.status === 'occupied') {
            return 'used-in-other-window';
        }
        if (device.status === 'used') {
            return 'used-in-other-window';
        }
        if (device.firmware === 'outdated') {
            return 'firmware-recommended';
        }
        return 'connected';
    }
    if (!device.available) {
        // deprecated
        return 'unavailable';
    }
    if (device.type === 'unacquired') {
        return 'unacquired';
    }
    if (device.type === 'unreadable') {
        return 'unreadable';
    }
    return 'unknown';
};

export const getStatusName = (deviceStatus: string): string => {
    switch (deviceStatus) {
        case 'connected':
            return 'Connected';
        case 'disconnected':
            return 'Disconnected';
        case 'bootloader':
            return 'Connected (bootloader mode)';
        case 'initialize':
            return 'Connected (not initialized)';
        case 'seedless':
            return 'Connected (seedless mode)';
        case 'firmware-required':
            return 'Connected (update required)';
        case 'firmware-recommended':
            return 'Connected (update recommended)';
        case 'used-in-other-window':
            return 'Used in other window';
        case 'unacquired':
            return 'Used in other window';
        case 'unavailable':
            return 'Unavailable';
        case 'unreadable':
            return 'Unreadable';
        default:
            return 'Status unknown';
    }
};

export const isWebUSB = (transport: Transport) => !!(transport.type && transport.type === 'webusb');

export const isDisabled = (
    selectedDevice: TrezorDevice,
    devices: Array<TrezorDevice>,
    transport: Transport
) => {
    if (isWebUSB(transport)) return false; // always enabled if webusb
    if (devices.length < 1) return true; // no devices
    if (devices.length === 1) {
        if (!selectedDevice.features) return true; // unacquired, unreadable
        if (selectedDevice.mode !== 'normal') return true; // bootloader, not initialized, seedless
        if (selectedDevice.firmware === 'required') return true;
    }
    return false; // default
};

export const isDeviceAccessible = (device: ?TrezorDevice): boolean => {
    if (!device || !device.features) return false;
    return device.mode === 'normal' && device.firmware !== 'required';
};

export const isSelectedDevice = (
    selected: ?TrezorDevice,
    device: ?(TrezorDevice | Device)
): boolean =>
    !!(
        selected &&
        device &&
        (selected.path === device.path && (device.ts && selected.instance === device.instance))
    );

export const getVersion = (device: TrezorDevice): string => {
    let version;
    if (device.features && device.features.major_version > 1) {
        version = 'T';
    } else {
        version = 'One';
    }
    return version;
};

export const getStatusColor = (deviceStatus: string): string => {
    switch (deviceStatus) {
        case 'connected':
            return colors.GREEN_PRIMARY;
        case 'disconnected':
            return colors.ERROR_PRIMARY;
        case 'bootloader':
        case 'initialize':
        case 'seedless':
        case 'firmware-recommended':
        case 'used-in-other-window':
        case 'unacquired':
            return colors.WARNING_PRIMARY;
        case 'firmware-required':
        case 'unavailable':
        case 'unreadable':
            return colors.ERROR_PRIMARY;
        default:
            return colors.TEXT_PRIMARY;
    }
};
