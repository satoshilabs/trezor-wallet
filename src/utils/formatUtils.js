/* @flow */

import BigNumber from 'bignumber.js';

export const toDecimalAmount = (amount: string | number, decimals: number): string => {
    try {
        const bAmount = new BigNumber(amount);
        // BigNumber() returns NaN on non-numeric string
        if (bAmount.isNaN()) {
            throw new Error('Amount is not a number');
        }
        return bAmount.div(10 ** decimals).toString(10);
    } catch (error) {
        return '0';
    }
};

export const fromDecimalAmount = (amount: string | number, decimals: number): string => {
    try {
        const bAmount = new BigNumber(amount);
        // BigNumber() returns NaN on non-numeric string
        if (bAmount.isNaN()) {
            throw new Error('Amount is not a number');
        }
        return bAmount.times(10 ** decimals).toString(10);
    } catch (error) {
        return '0';
    }
};
