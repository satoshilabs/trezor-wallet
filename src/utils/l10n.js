import { LANGUAGE } from 'config/variables';

export const getInitialLocale = (defaultLocale = 'en') => {
    const browserLocale = navigator.language.split('-')[0];
    if (LANGUAGE.some(e => e.code === browserLocale)) {
        // Array of supported languages contains the locale we're looking for
        return browserLocale;
    }
    // if browser lang is not supported return en as default locale
    return defaultLocale;
};