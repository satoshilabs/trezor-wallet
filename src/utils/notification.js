import colors from 'config/colors';
import icons from 'config/icons';

const getPrimaryColor = type => {
    let color;
    switch (type) {
        case 'info':
            color = colors.INFO_PRIMARY;
            break;
        case 'error':
            color = colors.ERROR_PRIMARY;
            break;
        case 'warning':
            color = colors.WARNING_PRIMARY;
            break;
        case 'success':
            color = colors.SUCCESS_PRIMARY;
            break;
        default:
            color = null;
    }

    return color;
};

const getSecondaryColor = type => {
    let color;
    switch (type) {
        case 'info':
            color = colors.INFO_SECONDARY;
            break;
        case 'error':
            color = colors.ERROR_SECONDARY;
            break;
        case 'warning':
            color = colors.WARNING_SECONDARY;
            break;
        case 'success':
            color = colors.SUCCESS_SECONDARY;
            break;
        default:
            color = null;
    }

    return color;
};

const getIcon = type => icons[type.toUpperCase()];

export { getPrimaryColor, getSecondaryColor, getIcon };
