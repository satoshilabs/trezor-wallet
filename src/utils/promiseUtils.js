/* @flow */


// import root from 'window-or-global';
// import Promise from 'es6-promise';

export async function resolveAfter(msec: number, value?: any): Promise<any> {
    await new Promise((resolve) => {
        //root.setTimeout(resolve, msec, value);
        window.setTimeout(resolve, msec, value);
    });
}