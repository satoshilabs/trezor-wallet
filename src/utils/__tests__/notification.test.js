import * as nUtils from 'utils/notification';

describe('device utils', () => {
    it('get status', () => {
        const types = [
            'info',
            'error',
            'warning',
            'success',
            'kdsjflds',
            '',
        ];

        types.forEach((type) => {
            expect(nUtils.getColor(type)).toMatchSnapshot();
        });
    });
    it('get icon', () => {
        const types = [
            'info',
            'error',
            'warning',
            'success',
            'kdsjflds',
            '',
        ];

        types.forEach((type) => {
            expect(nUtils.getIcon(type)).toMatchSnapshot();
        });
    });
});
