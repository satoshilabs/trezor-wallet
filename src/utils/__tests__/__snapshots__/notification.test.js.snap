// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`device utils get icon 1`] = `
Array [
  "M693.024 330.944c-99.968-99.936-262.080-99.936-362.048 0s-99.968 262.112 0 362.080c99.968 100 262.144 99.936 362.048 0 99.968-99.904 99.968-262.176 0-362.080zM507.904 300.192c27.008 0 48.992 21.984 48.992 49.088 0 27.296-21.984 49.472-48.992 49.472-27.264 0-49.536-22.176-49.536-49.472 0-27.552 21.728-49.088 49.536-49.088zM586.656 660.8c0 10.304-4.96 15.328-15.264 15.328h-126.464c-10.304 0-15.328-5.024-15.328-15.328v-32.256c0-10.304 5.024-15.264 15.328-15.264h23.36v-136.064h-23.872c-10.304 0-15.264-5.024-15.264-15.328v-32.224c0-10.304 4.96-15.264 15.264-15.264h88.288c10.304 0 15.264 4.96 15.264 15.264v183.648h23.424c10.304 0 15.264 4.96 15.264 15.264v32.224z",
]
`;

exports[`device utils get icon 2`] = `
Array [
  "M693.12 330.88c-46.317-46.267-110.276-74.88-180.919-74.88-141.385 0-256 114.615-256 256s114.615 256 256 256c70.642 0 134.602-28.613 180.921-74.882l-0.002 0.002c46.387-46.337 75.081-110.377 75.081-181.12s-28.694-134.783-75.079-181.118l-0.002-0.002zM494.080 344.32h53.12c16 0 18.24 9.28 18.24 14.72v10.24l-10.88 194.56c0 14.4-8 17.28-18.88 17.28h-28.16c-10.56 0-17.28-2.88-18.88-17.92l-10.88-193.92v-10.56c-1.28-4.8 2.24-14.080 16.32-14.080zM521.28 717.76c-0.095 0.001-0.207 0.001-0.319 0.001-27.747 0-50.24-22.493-50.24-50.24s22.493-50.24 50.24-50.24c27.747 0 50.24 22.493 50.24 50.24 0 0.112 0 0.224-0.001 0.336v-0.017c0 0 0 0.001 0 0.001 0 27.634-22.311 50.057-49.903 50.239h-0.017z",
]
`;

exports[`device utils get icon 3`] = `
Array [
  "M795.616 735.008l-264.896-465.44c-10.272-18.080-27.168-18.080-37.504 0l-264.864 465.44c-10.272 18.176-1.696 32.992 19.040 32.992h529.184c20.8 0 29.376-14.816 19.040-32.992zM549.76 673.12c0 10.464-8.48 18.976-18.912 18.976h-37.792c-10.336 0-18.912-8.512-18.912-18.976v-37.952c0-10.464 8.576-18.976 18.912-18.976h37.792c10.4 0 18.912 8.544 18.912 18.976v37.952zM549.76 559.264c0 10.464-8.48 18.976-18.912 18.976h-37.792c-10.336 0-18.912-8.512-18.912-18.976v-113.856c0-10.464 8.576-18.976 18.912-18.976h37.792c10.4 0 18.912 8.544 18.912 18.976v113.856z",
]
`;

exports[`device utils get icon 4`] = `
Array [
  "M692.8 313.92l-1.92-1.92c-6.246-7.057-15.326-11.484-25.44-11.484s-19.194 4.427-25.409 11.448l-0.031 0.036-196.48 224-3.84 1.6-3.84-1.92-48.64-57.28c-7.010-7.905-17.193-12.862-28.533-12.862-21.031 0-38.080 17.049-38.080 38.080 0 7.495 2.165 14.485 5.905 20.377l-0.092-0.155 100.8 148.16c5.391 8.036 14.386 13.292 24.618 13.44h8.662c17.251-0.146 32.385-9.075 41.163-22.529l0.117-0.191 195.2-296.32c4.473-6.632 7.141-14.803 7.141-23.597 0-11.162-4.297-21.32-11.326-28.911l0.025 0.028z",
]
`;

exports[`device utils get icon 5`] = `undefined`;

exports[`device utils get icon 6`] = `undefined`;

exports[`device utils get status 1`] = `"#1E7FF0"`;

exports[`device utils get status 2`] = `"#ED1212"`;

exports[`device utils get status 3`] = `"#EB8A00"`;

exports[`device utils get status 4`] = `"#01B757"`;

exports[`device utils get status 5`] = `null`;

exports[`device utils get status 6`] = `null`;
